### ISSUE TITLE
Dynamic Sort Function

### ISSUE THREAD

**Issue by tkohlman** [pantor/inja#109]
I'm trying to accomplish dynamic sorting on an array in the input json given an object key. I have arrays of objects such as { "id": 0, "name": "Name" }, and I would like the ability to process these objects by order of name or id, depending on the circumstances.

I have tried the following, but have not been able to make it work:

    environment.add_callback("sort_array", 2, [](inja::Arguments& args)
    {
        const nlohmann::json* array = args.at(0);
        auto key = args.at(1)->get<std::string>();

        if (!array->is_array())
        {
            return "";
        }

        std::sort(const_cast<nlohmann::json*>(array)->begin(), const_cast<nlohmann::json*>(array)->end(),
            [key](const nlohmann::json& object1, const nlohmann::json& object2) -> bool
            {
                const nlohmann::json& key1 = object1[key];
                const nlohmann::json& key2 = object2[key];

                if (key1.is_number() && key2.is_number())
                {
                    return key1.get<int>() < key2.get<int>();
                }
                else if (key1.is_string() && key2.is_string())
                {
                    return key1.get<std::string>() < key2.get<std::string>();
                }
                else
                {
                    return 0;
                }
            });

        return "";
    });

And then in my template:

`{{ sort_array(dataList, "id") }}`

If I iterate over the resulting list in the callback, the objects have been sorted. However, this does not result in modification of the input (i.e. the template is rendered in the unsorted list order).

This functionality would be nice in the form of a built-in function, or even in the form of extensible `void` custom functions (i.e. custom functions that modify the input instead of returning a string to render into the rendered template).

**Comment by tkohlman** [comment]
I was able to hack something together and get this working. The reason my `const_cast` hack failed is because the callback is passed a copy of the array. When the renderer resolves an input variable in a function call (i.e. the array for my purposes), it pushes a _copy_ of the input variable onto a stack.

`    std::vector<json> m_stack;`

Therefore, when I sort the copy, it has no effect on future iterations over the array.

I solved this issue by adding a second stack;

    std::vector<json*> m_tmp_stack;
    std::vector<json*> m_stack;

`m_stack` now stores pointers to json objects instead of json object copies. If a resolved value is from the input, then it is a simple pointer into the input. If the resolved value is a temporary value (e.g. the result of upper(), lower(), length(), etc), then it is allocated on the heap, pushed onto `m_tmp_stack` and pushed as a pointer onto `m_stack`; `m_stack` owns none of its contents and `m_tmp_stack` owns all of its contexts (and must delete them when popped).

I haven't benchmarked, but this implementation avoids a significant number of deep copies of the input json object, so it would likely boost the performance of the library.

I realize that this change is rather large, so feel free to reject it as an enhancement. The library is small enough that I was able to hack up what I need.
